homeassistant/components/music_assistant/media_player.py

```diff
@@ -0,0 +1,729 @@
+            return  # guard
+        if TYPE_CHECKING:
+            assert self.player.active_source is not None
+        if queue := self.mass.player_queues.get(self.player.active_source):
+            await self.mass.player_queues.transfer_queue(
+                mass_player_id, queue.queue_id, auto_play
+            )
+
+    async def async_browse_media(
+        self,
+        media_content_type: MediaType | str | None = None,
+        media_content_id: str | None = None,
+    ) -> BrowseMedia:
+        """Implement the websocket media browsing helper."""
+        return await media_source.async_browse_media(
+            self.hass,
+            media_content_id,
+            content_filter=lambda item: item.media_content_type.startswith("audio/"),
+        )
+
+    async def _get_item_by_name(
+        self,
+        name: str,
+        artist: str | None = None,
+        album: str | None = None,
+        media_type: str | None = None,
+    ) -> MediaItemType | None:
+        """Try to find a media item (such as a playlist) by name."""
+        # pylint: disable=too-many-nested-blocks
+        searchname = name.lower()
+
+        funcs: Sequence[Callable] = (
+            self.mass.music.get_library_playlists,
+            self.mass.music.get_library_radios,
+            self.mass.music.get_library_tracks,
+            self.mass.music.get_library_albums,
+            self.mass.music.get_library_artists,
+        )
+
+        library_functions = [
+            x for x in funcs if not media_type or media_type.lower() in x.__name__
+        ]
+        # prefer (exact) lookup in the library by name
+        for func in library_functions:
+            result = await func(search=searchname)
+            for item in result:
+                # handle optional artist filter
+                if (
+                    artist
+                    and (artists := getattr(item, "artists", None))
+                    and not any(x for x in artists if x.name.lower() == artist.lower())
+                ):
+                    continue
+                # handle optional album filter
+                if (
+                    album
+                    and (item_album := getattr(item, "album", None))
+                    and item_album.name.lower() != album.lower()
+                ):
+                    continue
+                if searchname == item.name.lower():
+                    return item  # type: ignore[no-any-return]
+        # nothing found in the library, fallback to global search
+        search_name = name
+        if album and artist:
+            search_name = f"{artist} - {album} - {name}"
+        elif album:
+            search_name = f"{album} - {name}"
+        elif artist:
+            search_name = f"{artist} - {name}"
+        result = await self.mass.music.search(
+            search_query=search_name,
+            media_types=[MediaType(media_type)] if media_type else MediaType.ALL,
+            limit=5,
+        )
```

Comment:
I kinda feel like this can be library code instead, since a search functionality is not HA specific

Reply:
This is a helper around the HA services but could be re-useable. 
Let's move it to the client library as a generic helper